saleor/plugins/manager.py

```diff
@@ -320,8 +320,15 @@ def list_payment_plugin_names(self, active_only: bool = False) -> List[tuple]:
             if payment_method in type(plugin).__dict__
         ]
 
-    def list_payment_gateways(self, active_only: bool = True) -> List[dict]:
+    def list_payment_gateways(
+        self, currency: Optional[str] = None, active_only: bool = True
+    ) -> List[dict]:
         payment_plugins = self.list_payment_plugin_names(active_only=active_only)
+        # if currency is given return only gateways which support given currency
+        if currency:
```

Comment:
Don't you think, that it would be better to append currencies to the response of the `list_payment_gateways`?
The `list_payment_plugin_names`  is used only in `list_payment_gateways`. Same as `__get_payment_config`. You can modify them to operate directly on the `plugin` not on the `name` - with `name` you need to fetch plugin every time when you want to get plugin data
Something like this:
```
    def get_payment_plugins(self, active_only: bool = False) -> Dict[str, "BasePlugin"]:
        payment_method = "process_payment"
        plugins = self.plugins
        if active_only:
            plugins = self.get_active_plugins()
        return {
            plugin.PLUGIN_ID: plugin
            for plugin in plugins
            if payment_method in type(plugin).__dict__
        }

    def list_payment_gateways(self, currency: None, active_only: bool = True) -> List[dict]:
        payment_plugins = self.get_payment_plugins(active_only=active_only)
        gateways = [
            {
                "id": plugin_id,
                "name": plugin.PLUGIN_NAME,
                "config": self.__get_payment_config(plugin),
                "currencies": self.__get_payment_currencies(plugin)
            }
            for plugin_id, plugin in payment_plugins
        ]
        if currency:
            return [gtw for gtw in gateways if currency in gtw["currencies"]]
        return gateways

    def __get_payment_currencies(self, gateway: "BasePlugin"):
        method_name = "get_payment_config"
        default_value: list = []
        config = self.__run_method_on_single_plugin(gateway, method_name, default_value)
        config_dict = {item["field"]:item['value'] for item in config}
        return config_dict.get('supported_currencies')


    def __get_payment_config(self, gateway: "BasePlugin") -> List[dict]:
        method_name = "get_payment_config"
        default_value: list = []
        return self.__run_method_on_single_plugin(gateway, method_name, default_value)
```

Reply:
I have a feeling that `if "process_payment" in type(plugin).__dict__` should really be an [abstract base class](https://docs.python.org/3/library/abc.html#abc.ABCMeta.__subclasshook__) so you could just use `isinstance(plugin, PaymentPlugin)`.

Reply:
We decided that the plugins need a bigger refactor, and we will leave it as it is for now.